[RFC] Hilbert space new API

PhilipVinc opened the pull request:
In general, in `netket` we do not _need_ to work with graphs, unless we are working with symmetries.
Especially to newcomers, but also to make the API more succinct, I think it would bee nice to support an usage like
```python
import netket
# 
>>> netket.hilbert.Spin(1/2)**4
Spin(s=1/2, graph=4)
>>> netket.operator.spin.sigmax(hi, 1)
<netket.operator._local_operator.LocalOperator object at 0x10f84fcd0>
```

Where basically we can define the local Hilbert space (which implicitly it will be defined on an Edgeless graph with 1 node) and then tensor-exponentiate it to define it on N sites.

Under the hood, `pow` on an Hilbert space is doing `pow` on the graph, which corresponds to a disjoint union (or direct sum) of the two graphs, which I believe to be the only sensible operation possible in this case.

Of course, this mainly makes sense when working with Edgeless graphs.

--

To make this kind of api work, we need to switch two arguments: the Spin rational. (1//2, 1, 3//2...) should become the first argument, so that if the graph (now second argument) is unspecified, we can assume it to be an `edgeless(1)` graph.

I'd propose to change all hilbert spaces to such api.

--

In the future, this should allow us to also define multiplication among hilbert spaces (technically \otimes, but python let's us only use `*`) and compose non-homogeneeous hilbert spaces together. But that's for a future PR, as it would require some work on the samplers.

If you agree with this new api, then I will also convert the other hilbert spaces (bosons, qubits, and local_operator).

femtobit commented:
Thanks for bringing this up, @PhilipVinc.

I am wondering (and I think we've had this discussion before for 2.0): Does the Hilbert space need to know about the graph at all? At least for the basic `Spin`, `Qubit`, and `Boson` spaces we have, that does not appear to be the case.

The graph is only needed further down the line for defining 
 1. some (but not all) operators 
 2. some (but not all) machines
 3. some (but not all) samplers.

Therefore, do we need our Hilbert spaces to depend on the graph at all? Would it not be cleaner to have an interface like:
```python
N = 16
hi = Spin(1/2, size=N)

op1 = sigmax(hi, [0])
op2 = Heisenberg(hi, Chain(N), J=1.0) # needs graph for coupling terms

ma = RbmSpin(hi, alpha=6) # does _not_ need graph (without imposed symmetries)
sampler1 = MetropolisLocal(ma)
sampler2 = MetrpolisExchange(ma, Chain(N)) # needs graph to exchange neighbors
```

Users need to make sure their graph and Hilbert space have the same size. Also, there has to be a consistent relation between "site" meaning "index in spin configuration" and "node on the graph".
At least the first point is easy to validate. I'm unsure whether the second part can be a problem.

On the other hand, the graph is not as universal as our current design makes it seem anyway. It is quite conceivable that one would want a different graph structure for a sampler (or, e.g., ConvNet machine) than for defining the Hamiltonian.

femtobit commented:
Interface wise this amounts to more or less the same as you suggestion, I think, but without the implicit `Edgeless` graph where no graph structure is actually needed.

gcarleo commented:
I agree with both of you actually (the only think I don't like too much though is setting ```graph=n``` to specify a edgeless graph of n sites... 
I also would be in favour of completely removing the graph from the Hilbert, but we have to be careful to see if this generates any problem upstream

femtobit commented:
@gcarleo Yes, if we remove the graph dependence from Hilbert, we can pass `Spin(..., N=16)` instead of `graph=16` or similar workarounds. Or, if the resulting object is the same, `Spin(...)**16` - although, while I have to admit it looks cool, I'd personally prefer to only fall back on the product/power notation for joining more complex spaces.

In any case, getting rid of the graph in Hilbert would finally make it a lot easier to define something like 
```python
hi = Spin(1/2, N=16) * Boson(N=1)
```

gcarleo commented:
I think that a quick and viable solution to implement now is just to set ```graph=None``` by default in the constructors of Hilbert and have an integer argument ```N``` . In this way we don't need to change very much, but just accept that hilbert can be constructed without a graph. Then the other classes can check if hilbert has a graph or not and decide whether to use it or not independently...

PhilipVinc commented:
@femtobit Thanks for the feedback. Indeed I agreee with you.

I have updated the PR in order to completely remove the graph from hilbert spaces, and added a new requireed argument to some hamiltonians (notably, Ising and other GraphOperators). Same is said for Exchange Sampler.

Please let me know what you think

PhilipVinc commented:
As for using `**N` or `pow` to indicate the tensor product of an hilbert space time itself N times...
I personally like it (it's the same as you find in QuantumOptics.jl and QuTip), and its meaning cannot be equivocated, therefore I'd lobby to keep it.

PhilipVinc commented:
By the way, now machines that accept `symmetry=True/False` take an argument `automorphisms` that can either be the list of automorphisms desired, or a graph from which they are computed.

VolodyaCO commented:
Is it now possible to use heterogeneous Hilbert spaces? (Spin and boson for
instance) as @gcarleo mentioned? Or not yet? Do the samplers still need to
be adjusted for this to work?

On Wed, 9 Dec 2020, 06:48 Filippo Vicentini, <notifications@github.com>
wrote:

> By the way, now machines that accept symmetry=True/False take an argument
> automorphisms that can either be the list of automorphisms desired, or a
> graph from which they are computed.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/netket/netket/pull/520#issuecomment-741721040>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHQJA722FTRBFNO4UNQFIUDST5PXPANCNFSM4UR34N6A>
> .
>

-- 
*Aviso legal:* El contenido de este mensaje y los archivos adjuntos son 
confidenciales y de uso exclusivo de la Universidad Nacional de Colombia. 
Se encuentran dirigidos sólo para el uso del destinatario al cual van 
enviados. La reproducción, lectura y/o copia se encuentran prohibidas a 
cualquier persona diferente a este y puede ser ilegal. Si usted lo ha 
recibido por error, infórmenos y elimínelo de su correo. Los Datos 
Personales serán tratados conforme a la Ley 1581 de 2012 y a nuestra 
Política de Datos Personales que podrá consultar en la página web 
www.unal.edu.co <http://www.unal.edu.co/>.* *Las opiniones, informaciones, 
conclusiones y cualquier otro tipo de dato contenido en este correo 
electrónico, no relacionados con la actividad de la Universidad Nacional de 
Colombia, se entenderá como personales y de ninguna manera son avaladas por 
la Universidad.

femtobit commented on Examples/BoseHubbard1d/bosehubbard1d.py:
```diff
@@ -18,13 +18,13 @@
 
 # Bose Hubbard Hamiltonian
-ha = nk.operator.BoseHubbard(U=4.0, hilbert=hi)
+ha = nk.operator.BoseHubbard(hilbert=hi, graph=g.n_nodes, U=4.0)
```
This should pass the graph, not `n_nodes`, right? (Similarly in several other examples.)
```suggestion
ha = nk.operator.BoseHubbard(hilbert=hi, graph=g, U=4.0)
```

PhilipVinc replied:
yes indeed. I think I messed up a search and replace

gcarleo commented on Examples/AKLT/AKLT_ed.py:
```diff
@@ -39,13 +39,14 @@
 
 # AKLT model Hamiltonian as graph
-ha = nk.operator.GraphOperator(hilbert=hi, bond_ops=[P2_AKLT.tolist()])
+ha = nk.operator.GraphOperator(hilbert=hi, graph=g.n_nodes, bond_ops=[P2_AKLT.tolist()])
```
```suggestion
ha = nk.operator.GraphOperator(hilbert=hi, graph=g.n_nodes, bond_ops=[P2_AKLT.tolist()])
```

gcarleo commented on Examples/BoseHubbard1d/bosehubbard1d.py:
```diff
@@ -18,13 +18,13 @@
 
 # RBM Machine with one-hot encoding, real parameters, and symmetries
-ma = nk.machine.RbmMultiVal(hilbert=hi, alpha=1, dtype=float, symmetry=True)
+ma = nk.machine.RbmMultiVal(hilbert=hi, alpha=1, dtype=float, automorphisms=g)
```
here you should pass automorphisms=g.automorphisms ?

PhilipVinc replied:
Can be both

femtobit commented on netket/graph/graph.py:
```diff
@@ -97,6 +97,19 @@ def __repr__(self):
             str(type(self)).split(".")[-1][:-2], self.n_nodes
         )
 
+    def __pow__(self, exponent):
+        if not isinstance(exponent, int):
+            raise ValueError(
+                "Exponent {} not valid: only integers are supported".format(exponent)
+            )
+
+        base = self
+        g = self
+        for i in range(1, exponent):
+            g = disjoint_union(g, base)
+
+        return g
+
```
[Graph power](https://en.wikipedia.org/wiki/Graph_power) has a separate meaning in graph theory and, furthermore, the disjoint union implemented here is more like a sum than a product (which is also how it is typically denoted). All the notions of [graph product](https://en.wikipedia.org/wiki/Graph_product) add edges between the factor graphs.

Therefore, I think using `__pow__` is confusing here. In particular, `|disjoint_union(G, H)| = |G| + |H|` (where `|·|` denotes `n_nodes`), so we would have `|G**N| = N * |G|`.

Therefore, (and since this operation now is also unrelated to `hilbert**N`) I'd leave it out or - if we still want to have this method - rename it, for example to `G.disjoint_copies(N)`.

PhilipVinc replied:
Indeed.

femtobit commented on netket/hilbert/boson.py:
```diff
@@ -8,27 +8,25 @@
         Args:
-           graph: Graph representation of sites.
            n_max: Maximum occupation for a site (inclusive). If None, the local occupation
                   number is unbounded.
-           n_bosons: Constraint for the number of bosons. If None, no constraint
+           N: Number of bosonic sites.
```
n_max and N are also optional, not just n_bosons as the comments indicate.

gcarleo replied:
how can N be optional?

femtobit replied:
It is `N=1` by default in the code at the moment.

femtobit commented on netket/hilbert/boson.py:
```diff
@@ -8,27 +8,25 @@
     r"""Hilbert space obtained as tensor product of local bosonic states."""
 
-    def __init__(self, graph, n_max=None, n_bosons=None):
+    def __init__(self, n_max=None, N=1, n_bosons=None):
```
```suggestion
    def __init__(self, *, n_max=None, N=1, n_bosons=None):
```
May prevent mixups.

femtobit commented on netket/hilbert/spin.py:
```diff
@@ -123,8 +121,16 @@ def _check_total_sz(self, total_sz, size):
+        if self._total_sz is None:
+            total_sz = None
+        else:
+            total_sz * n
```
There is an `=` missing. (If the tests pass, we should probably add a test for power with constraint, as this line is apparently not covered.)

femtobit commented on netket/machine/jastrow.py:
```diff
@@ -45,8 +46,8 @@ def __init__(
             use_visible_bias (bool): Whether to use the visible bias a_i.
-            symmetry (optional): if True, the automorphisms in hilbert.graph.automorphisms are used.
-                                 A custom matrix of automorphisms can also be passed instead.
+            symmetry (optional): Can be a graph or a custom matrix of automorphisms can
```
```suggestion
            automorphisms (optional): Can be a graph or a custom matrix of automorphisms can
```

femtobit commented:
There are some bugs related to the Hilbert space power:
```python
>>> hi = Boson()
>>> hi
Boson(n_max=INT_MAX; N=1)
>>> hi = Boson()**2 # does not fail, but ...
>>> hi              # ...this does
TypeError: object of type 'NoneType' has no len()
```
This is because `__pow__` silently changes the type and `CustomHilbert` cannot handle unconstrained bosons:
```python
>>> type(Boson()), type(Boson()**2)
(netket.hilbert.boson.Boson, netket.hilbert.custom_hilbert.CustomHilbert)
```

gcarleo commented on Examples/CustomGraph/custom_graph.py:
```diff
@@ -22,10 +22,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```

gcarleo commented on Examples/CustomHamiltonian/custom_hamiltonian.py:
```diff
@@ -20,7 +20,7 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
```

gcarleo commented on Examples/CustomMachine/ising1d.py:
```diff
@@ -19,10 +19,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
```

gcarleo commented on Examples/CustomSampler/customsampler_heisenberg1d.py:
```diff
@@ -21,13 +21,13 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/CustomSampler/exchange_kernel.py:
```diff
@@ -6,7 +6,7 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/DissipativeIsing1d/ising1d.py:
```diff
@@ -25,7 +25,7 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/EarlyStopping/heisenberg1d.py:
```diff
@@ -21,13 +21,13 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/ExactDiag/dissipative_ising1.py:
```diff
@@ -23,7 +23,7 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/ExactDiag/heisenberg1.py:
```diff
@@ -19,7 +19,7 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/GraphOperator/Ising/ising.py:
```diff
@@ -29,7 +29,7 @@
 
 # Custom Hilbert Space
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/GraphOperator/J1J2/j1j2.py:
```diff
@@ -50,7 +50,7 @@
 
 # Spin based Hilbert Space
-hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, graph=g)
+hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, total_sz=0.0, N=g.n_nodes)
```

gcarleo commented on Examples/Heisenberg1d/heisenberg1d.py:
```diff
@@ -19,13 +19,13 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/Heisenberg1d/heisenberg1d_conv.py:
```diff
@@ -20,7 +20,7 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/Heisenberg1d/heisenberg1d_ffnn.py:
```diff
@@ -20,7 +20,7 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/Heisenberg1d/heisenberg1d_jastrow.py:
```diff
@@ -19,13 +19,13 @@
 # Hilbert space of spins on the graph
 # with total Sz equal to 0
-hi = nk.hilbert.Spin(s=0.5, graph=g, total_sz=0)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes, total_sz=0)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2,N=g.n_nodes, total_sz=0)
```

gcarleo commented on Examples/Ising1d/ising1d.py:
```diff
@@ -18,10 +18,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/Ising1d/ising1d.py:
```diff
@@ -18,10 +18,10 @@
 
 # Ising spin hamiltonian
-ha = nk.operator.Ising(h=1.0, hilbert=hi)
+ha = nk.operator.Ising(h=1.0, hilbert=hi, graph=g.n_nodes)
```
```suggestion
ha = nk.operator.Ising(h=1.0, hilbert=hi, graph=g)
```

gcarleo commented on Examples/Ising1d/ising1d_jastrow.py:
```diff
@@ -18,10 +18,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/Ising1d/torch_rbm.py:
```diff
@@ -22,10 +22,10 @@ def forward(self, x):
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/J1J2/j1j2.py:
```diff
@@ -44,15 +44,15 @@
 
 # Spin based Hilbert Space
-hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, graph=g)
+hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, N=g.n_nodse)
```
```suggestion
hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, N=g.n_nodes)
```

gcarleo commented on Examples/RealMachines/j1j2.py:
```diff
@@ -44,7 +44,7 @@
 
 # Spin based Hilbert Space
-hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, graph=g)
+hi = nk.hilbert.Spin(s=0.5, total_sz=0.0, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, total_sz=0.0, N=g.n_nodes)
```

gcarleo commented on Examples/RealMachines/j1j2.py:
```diff
@@ -56,7 +56,7 @@
 
 # Sampler
-sa = nk.sampler.MetropolisExchange(machine=ma, graph=g)
+sa = nk.sampler.MetropolisExchange(machine=ma, graph=g.n_nodes)
```
```suggestion
sa = nk.sampler.MetropolisExchange(machine=ma, N=g.n_nodes)
```

gcarleo commented on Examples/RealMachines/rbm_phase.py:
```diff
@@ -18,10 +18,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/RealMachines/rbm_real.py:
```diff
@@ -18,10 +18,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/Supervised/Ising/ed.py:
```diff
@@ -21,10 +21,10 @@ def load_ed_data(L):
 
     # Hilbert space of spins on the graph
-    hi = nk.hilbert.Spin(s=0.5, graph=g)
+    hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
    hi = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
```

gcarleo commented on Examples/Supervised/J1J2/ed.py:
```diff
@@ -44,7 +44,7 @@ def load_ed_data(L, J2=0.4):
 
     # Spin based Hilbert Space
-    hi = nk.hilbert.Spin(s=0.5, graph=g)
+    hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
    hi = nk.hilbert.Spin(s=0.5, N=g.n_nodes)
```

gcarleo commented on Examples/Supervised/J1J2/load_data.py:
```diff
@@ -24,7 +24,7 @@ def load(path_to_samples, path_to_targets):
     # TODO remove Hypercube here and put customgraph
     g = gr.Hypercube(length=len(tsamples[0]), n_dim=1)
-    hi = hs.Qubit(graph=g)
+    hi = hs.Qubit(graph=g.n_nodes)
```
```suggestion
    hi = hs.Qubit(N=g.n_nodes)
```

gcarleo reviewed (changes requested):
half way through

gcarleo commented on Examples/Tensorboard/ising1d.py:
```diff
@@ -22,10 +22,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

gcarleo commented on Examples/Torch/torch_machine.py:
```diff
@@ -20,10 +20,10 @@
 
 # Hilbert space of spins on the graph
-hi = nk.hilbert.Spin(s=0.5, graph=g)
+hi = nk.hilbert.Spin(s=0.5, graph=g.n_nodes)
```
```suggestion
hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)
```

PhilipVinc commented:
@femtobit Yes, I noticed now that Boson is a subclass of CustomHilbert and this was causing some complications... I defined `__pow__` for every hilbert type now so it should be fine, plus tests.

@gcarleo I fixed all examples

gcarleo commented:
@PhilipVinc what does it happen when you multiply two bosons hilberts (a, and b, say) with different number of total bosons? is the resulting class carrying n_bosons=n_bosons(a)+n_bosons(b) ?

femtobit commented on netket/hilbert/custom_hilbert.py:
```diff
@@ -206,6 +213,15 @@ def _to_constrained_numbers_kernel(do_constraints, bare_numbers, numbers):
                 )
             return found
 
+    def __pow__(self, n):
+        if self._constraints is not None:
+            raise ValueError(
+                """Cannot exponentiate a CustomHilbret with a constraint. 
```
```suggestion
    def __pow__(self, n):
        if self._constraints is not None:
            raise NotImplementedError(
                """Cannot exponentiate a CustomHilbert with constraints. 
```

PhilipVinc commented:
>@PhilipVinc what does it happen when you multiply two bosons hilberts (a, and b, say) with different number of total bosons? is the resulting class carrying n_bosons=n_bosons(a)+n_bosons(b) ?

This operation is currently not implemented. The only operation implemented is `exponentiation` of hilbert spaces.
In that case I multiply the number of sites, and, if present, total number of bosons, by the _exponent_

gcarleo reviewed (approved):
Looks good now!

femtobit commented on netket/hilbert/spin.py:
```diff
@@ -68,7 +68,7 @@ def random_state(self, out=None, rgen=None):
            >>> import netket as nk
            >>> import numpy as np
-           >>> hi = nk.hilbert.Boson(n_max=3, graph=nk.graph.Hypercube(length=5, n_dim=1))
+           >>> hi = nk.hilbert.Boson(n_max=3, N=4)
```
```suggestion
           >>> hi = nk.hilbert.Spin(N=4)
```

femtobit reviewed (commented):
Looks very good to me, just one minor thing I've noticed.

gcarleo commented:
I am having "slight" second thoughts... shouldn't we just allow ALSO for the extra argument graph=None so that people don't need to rewrite ALL their codes?

PhilipVinc commented:
ops. I just merged.
What you are proposing is essentially a deprecation warning.
I'm fine with it. We can remove it in v3.1.

I'll add it now in another PR

PhilipVinc commented:
Wait, @gcarleo, do you want it to accept graph (and take the number of nodes from it) but still people will need to pass graph to Ising/GraphOprator, or do you want hilbert to take a graph and store it?

gcarleo commented:
I think we can just accept it in the constructor to automatically deduce N, but never store it...
